lib/util/Ordering.ts

```diff
@@ -1,29 +1,57 @@
 import type * as RDF from '@rdfjs/types';
 import * as LRUCache from 'lru-cache';
-import type * as T from '../expressions/Term';
+import * as T from '../expressions/Term';
 import { TermTransformer } from '../transformers/TermTransformer';
 import type { ISuperTypeProvider, SuperTypeCallback, TypeCache } from './TypeHandling';
 
 // Determine the relative numerical order of the two given terms.
 export function orderTypes(litA: RDF.Term | undefined, litB: RDF.Term | undefined, isAscending: boolean,
   typeDiscoveryCallback?: SuperTypeCallback, typeCache?: TypeCache, enableExtendedXSDTypes?: boolean): -1 | 0 | 1 {
+  if (litA === undefined || litB === undefined || litA.equals(litB)) {
+    return 0;
+  }
+  return isLowerThan(litA, litB, typeDiscoveryCallback, typeCache, enableExtendedXSDTypes) === isAscending ? -1 : 1;
+}
+
+function isLowerThan(litA: RDF.Term, litB: RDF.Term,
+  typeDiscoveryCallback?: SuperTypeCallback, typeCache?: TypeCache, enableExtendedXSDTypes?: boolean): boolean {
   const openWorldType: ISuperTypeProvider = {
     discoverer: typeDiscoveryCallback || (() => 'term'),
     cache: typeCache || new LRUCache(),
   };
   const termTransformer = new TermTransformer(openWorldType, enableExtendedXSDTypes || false);
-  if (litA && litA.termType === 'Literal' && litB && litB.termType === 'Literal') {
+  if (litA.termType === 'Literal' && litB.termType === 'Literal') {
     const myLitA = termTransformer.transformLiteral(litA);
     const myLitB = termTransformer.transformLiteral(litB);
-    return order(myLitA, myLitB, isAscending);
+    const typeA = _TYPE_BUCKETS.get(myLitA.constructor) || 'unknown';
+    const typeB = _TYPE_BUCKETS.get(myLitB.constructor) || 'unknown';
+    if (typeA !== typeB) {
+      return typeA < typeB;
+    }
+    switch (typeA) {
+      case 'boolean':
+      case 'dateTime':
+      case 'number':
+      case 'string':
+        return myLitA.typedValue < myLitB.typedValue;
+      case 'langString':
+        return myLitA.typedValue < myLitB.typedValue ||
+            (myLitA.typedValue === myLitB.typedValue && (myLitA.language || '') < (myLitB.language || ''));
+      default:
+        return myLitA.dataType < myLitB.dataType ||
+            (myLitA.dataType === myLitB.dataType && myLitA.str() < myLitB.str());
+    }
   }
-  return 0;
+  // Order is not defined yet for other terms.
+  return true;
 }
 
-// Effective ordering
-export function order(orderA: T.Literal<any>, orderB: T.Literal<any>, isAscending: boolean): -1 | 0 | 1 {
-  if (orderA.typedValue === orderB.typedValue) {
-    return 0;
-  }
-  return orderA.typedValue > orderB.typedValue === isAscending ? 1 : -1;
-}
+const _TYPE_BUCKETS = new Map<any, string>();
+_TYPE_BUCKETS.set(T.BooleanLiteral, 'boolean');
```

Comment:
I would recommend indexing these buckets by `dataType` (there should be constants defined for these somewhere), to reducing coupling with our literal class implementation.

Reply:
Also, instead of defining the type buckets manually here, I'm wondering if we can't reuse some definitions from elsewhere for that.
For example: https://github.com/comunica/sparqlee/blob/a4eaeacb42f384457f9cee64a5b258efc2ff8747/lib/util/TypeHandling.ts#L19

Perhaps @jitsedesmet or @wschella have some thoughts on this.

Reply:
Thank you! Indeed, `mainSparqlType` seems the perfect function for this use case.

Reply:
I am fraid the coupling wih literal class is slightly unavoidable because the ordering implement for each kind of literals depends on how the literals are parsed. A cleaner option might be to rely on the `Literal.mainSparqlType` attribute instead of the class names.

`mainSparqlType` does not easily solve the problem because it would require to run the `typeDiscoveryCallback` before to make sure we use the normalized datatype and not the base one.

Reply:
> A cleaner option might be to rely on the Literal.mainSparqlType attribute instead of the class names.

Sounds good to me!

> mainSparqlType does not easily solve the problem because it would require to run the typeDiscoveryCallback before to make sure we use the normalized datatype and not the base one.

Not sure I fully understand. AFAICS, `typeDiscoveryCallback` has already been called. And if it isn't, can't we invoke it directly here?

Reply:
Oh, the scheme for this was /very/ easy to find since it is the only document in the `docs` folder.  https://github.com/comunica/sparqlee/blob/master/docs/type-scheme.svg

Reply:
Thank you so much for the review and the suggestion. I just pushed an other attempt 8e9e86c07e53f08623b2e02987dd81ea9599bb40 using the `Literal` `mainSparqlType` field. This makes easy to type check the code and see if it's exhaustive. This way if a new `MainSparqlType` is added the type checker will require to update the ordering code to support it.

The internal sparqlee types are more fine grained and I believe don't need this level of details for the ordering imho. Using directly the `mainSparqlType` also allows to make use of the `typeDiscoveryCallback` evaluation done during literal transformation.

Thank you again for your suggestion. I hope my new proposal is moving things forward.

Reply:
Oh I see now, the stronger/ experimental type system makes sure you still have the correct `mainSparlType` even if the type has to be discovered. The type handling seems good now. As for implementation, that's for Ruben to decide. :slightly_smiling_face: